### DISCUSSION TITLE
Questions about the radius_graph

### DISCUSSION THREAD

**Discussion by StefanIsSmart** [pyg-team/pytorch_geometric#8764] [General]
The radius_graph in pyg could calculate the point that is in a distance (r) right?
However, I am wondering how to calculate the distance in implementation.
The Euclidean Distance？

**Comment by rusty1s** [comment]
Yes, you can just do
```python
edge_index = radius_graph(x, r=...)
distance = (x[edge_index[0]] - x[edge_index[1]]).norm(dim=-1)
```
to get the final distance.

**Reply by StefanIsSmart** [reply]
My meaning is :
edge_index = radius_graph(x, r=...)
If the x is the coordinates of mol. (XYZ), r is the cutoff like 3 Angstrom. 
So the function will build a link between the nodes that their Euclidean Distance in 3 Angstrom?

**Reply by StefanIsSmart** [reply]
In fact, before asking this question, I always thought this function could return the paired nodes that their Euclidean Distance in 3 Angstrom. But one time I tried to set the r as 1e10 (the x is the xyz coordinates of a real molecule.), which means the graph would be fully connected, but the result was not as I expected.

**Reply by rusty1s** [reply]
Yes, `radius_graph` is based on Euclidean distance. Currently, there is an upper limit on the number of neighbors returned, denoted by the `max_num_neighbors` argument in `radius_graph`.

**Reply by StefanIsSmart** [reply]
Thank you !
So by now, my understanding is:
The radius_graph and the knn_graph are all calculating the distance based on Euclidean distance.
The difference point is the knn_graph will return the nearest k nodes, and the raduis_graph will return all the nodes in a `distance cutoff`.
If I set the `distance cutoff` as positive infinite and set the `max_num_neighbors` as k, the radius graph will == the knn_graph. 
Right?

**Reply by rusty1s** [reply]
Yes, this is mostly correct. However, distance cutoff==infinity + `max_num_neighbors=k` is not guaranteed to return the closest k neighbors. It will randomly pick k neighbors that fall into the distance cutoff (at least on GPU).

**Reply by StefanIsSmart** [reply]
I see, thank you! 
I am a new fish in PyG, it is very helpful for me!
